### DISCUSSION TITLE
Tomcat Custom Login Module

### DISCUSSION THREAD

**Discussion by ernstba24** [hawtio/hawtio#3378] [Q&A]
Are there any examples on how to configure a custom login module for hawtio in a tomcat container. The official docs only provide configuration steps for jetty servers.  Havins some issue on what's expected in cong/login.conf file in order for hawtio to pick up the login module

**Comment by grgrzybek** [comment]
https://hawt.io/docs/security.html#_using_different_login_modules is indeed a bit short and refers to Jetty.
But mind that JAAS is container agnostic and relevant configuration can be summarized with:

1. Set system property `java.security.auth.login.config` - for example add `export JAVA_OPTS='-Djava.security.auth.login.config=$CATALINA_BASE/conf/login.conf'` in Tomcat's `bin/setenv.sh`
2. Create `login.conf` (configured above) with the content [as configured in Appendix B: JAAS Login Configuration File](https://docs.oracle.com/en/java/javase/11/security/appendix-b-jaas-login-configuration-file.html#GUID-9713B697-EFED-49A1-9E15-8039AD04458B), for example (standard LDAP login module):
```
my-realm {
    com.sun.security.auth.module.LdapLoginModule REQUIRED
    userProvider="ldap://localhost:389"
    authIdentity="uid={USERNAME},ou=users,dc=example,dc=com"
    useSSL=false
    debug=true;
};
```
3. Ensure that `my-realm` realm is used by Hawtio with `-Dhawtio.realm=my-realm` (again in Tomcat, add it in `bin/setenv.sh`
4. If the login module's class is not standard, ensure that the JAR providing this class is available - add it to `lib` folder in Tomcat

This should be enough to make it work. If it doesn't, please show your configuration.
